File: Data Structures & Algorithms/palindrome-partitioning/submission-1.java
class Solution {
    List<List<String>> ans = new ArrayList<>();
    String s;
    char[] cs;
    public List<List<String>> partition(String s) {
        this.s = s;
        cs = s.toCharArray();
        backtrack(0, new ArrayList<>());
        return ans;
    }
    void backtrack(int start, List<String> cur) {
        if (start == s.length()) {
            ans.add(new ArrayList<>(cur));
            return;
        }
        for (int end = start; end < s.length(); end ++) {
            if (isPalindrome(start, end)) {
                cur.add(s.substring(start, end+1));
                backtrack(end+1, cur);
                cur.remove(cur.size()-1);
            }
        }
    }
    private boolean isPalindrome(int start, int end) {
        int i = start, j = end;
        while (i <= j) {
            if (cs[i] != cs[j]) return false;
            i++;j--;
        }
        return true;
    }
}

File: Data Structures & Algorithms/is-palindrome/submission-1.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        
                
        def isAlphanumeric(c):
            return (ord(c)>=ord('a') and ord(c)<=ord('z')) or (ord(c)>=ord('A') and ord(c)<=ord('Z')) or (ord(c)>=ord('0') and ord(c)<=ord('9'))
        
        i = 0
        j = len(s) - 1
        while i < j:
            while i < len(s) and not isAlphanumeric(s[i]):
                i +=1
            while j >= 0 and not isAlphanumeric(s[j]):
                j -=1
            if i<len(s) and j>=0 and s[i].lower() != s[j].lower():
                return False
            i += 1
            j -= 1
        return True